backend/api/task.proto

```diff
@@ -0,0 +1,110 @@
+// Copyright 2021 The Kubeflow Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+option go_package = "github.com/kubeflow/pipelines/backend/api/go_client";
+package api;
+
+import "google/protobuf/empty.proto";
+import "google/api/annotations.proto";
+import "google/protobuf/timestamp.proto";
+import "backend/api/resource_reference.proto";
+import "protoc-gen-swagger/options/annotations.proto";
+
+message Task {
+  // Output. Unique run ID. Generated by API server.
+  string id = 1;
+
+  // Required input field. The Namespace to which this pipeline task belongs.
+  string namespace = 2;
+
+  // Required input field. The PipelineName to which this pipeline task belongs.
+  string pipelineName = 3;
+
+  // Required input field.The ID of the PipelineRun that the PipelineTask belongs to.
+  string runId = 4;
+
+  // Required input field. The ID of the MLMD execution associated with the PipelineTask.
+  string mlmdExecutionID = 5;
```

Comment:
note, the common practice in KFP for referring to other related resources is to use the resource reference object. See https://github.com/kubeflow/pipelines/blob/3594e898038599b7c69d9de0071585e87d65a623/backend/api/run.proto#L234-L237.

However, I don't quite like it now, because it makes the API interface much harder to use and understand. Were you intentionally not following this pattern? what were your rationale?

Reply:
I took a look at the codebase. Currently the resourceReferenes in db layer only tracks the relationship between the resources recorded in the backend db like `Experiment`, `Job` and `Run`. For other relationship like the relationship between "namespace" and "experiment", it won't records it in resourceReferences. I think we don't need the relationship between "mlmdExecutionID" and "pipelineTask" in `resourceReferences` table like what we do now for "namespace" and "experiment". I think "mlmdExecutionID" is just a property of the "pipelineTask" and "mlmdExecutionID" is stored in MLMD instead of in KFP backend DB.

For the ResourceReference in KFP, I think the storage layer just follows the interface in that API layer for the resources recorded in KFP backend DB. For the API layer, I think resource_references make it the api layer more flexible though a little bit difficult to understand.

Reply:
>  I think it's a trade off between the flexibility of the api(the stability of the interface of api layer) and the easability of the api.

That's a good summary of the tradeoff, but I have a different perception.

1. Interface data structure doesn't change when adding new relationships does not imply API is stable. API also includes what data is expected, if expected data changes, it's also an API change, e.g. adding a required relationship or removing a past supported relationship